File: src/screen_probe.js
// ZeroClaw Hybrid Programmatic Grounding — Tier 2: JXA System Events Probe
// Run via: osascript -l JavaScript screen_probe.js
//
// Uses System Events to traverse the frontmost application's UI element tree.
// Slower than Swift AXAPI (uses Apple Events IPC per property) but always available.

ObjC.import("Foundation");
ObjC.import("AppKit");

var elementId = 0;
var MAX_DEPTH = 10;
var MAX_ELEMENTS = 300;

var interactableRoles = {
  AXButton: true,
  AXTextField: true,
  AXTextArea: true,
  AXCheckBox: true,
  AXRadioButton: true,
  AXPopUpButton: true,
  AXComboBox: true,
  AXSlider: true,
  AXLink: true,
  AXMenuItem: true,
  AXMenuBarItem: true,
  AXTab: true,
  AXDisclosureTriangle: true,
  AXIncrementor: true,
  AXCell: true,
  AXRow: true,
  AXSwitch: true,
  AXToggle: true,
};

var skipRoles = {
  AXGroup: true,
  AXSplitGroup: true,
  AXSplitter: true,
  AXScrollArea: true,
  AXLayoutArea: true,
  AXLayoutItem: true,
  AXUnknown: true,
};

function safeGet(fn) {
  try {
    var v = fn();
    return v === null || v === undefined ? null : v;
  } catch (e) {
    return null;
  }
}

function traverseElement(el, depth, elements) {
  if (depth >= MAX_DEPTH || elements.length >= MAX_ELEMENTS) return;

  var role = safeGet(function () {
    return el.role();
  });
  if (!role) return;

  var name = safeGet(function () {
    return el.name();
  });
  if (!name)
    name = safeGet(function () {
      return el.description();
    });

  var value = safeGet(function () {
    var v = el.value();
    if (typeof v === "string") {
      return v.length > 200 ? v.substring(0, 197) + "..." : v;
    }
    if (typeof v === "number") return String(v);
    return null;
  });

  var bbox = null;
  try {
    var pos = el.position();
    var sz = el.size();
    if (pos && sz && sz[0] > 0 && sz[1] > 0) {
      bbox = { x: pos[0], y: pos[1], w: sz[0], h: sz[1] };
    }
  } catch (e) {}

  var interactable = !!interactableRoles[role];
  // JXA/System Events assigns generic names ("group", "scroll area") to containers.
  // These are noise — treat them as no name.
  var genericNames = {
    group: 1,
    "split group": 1,
    "scroll area": 1,
    "layout area": 1,
    "layout item": 1,
    splitter: 1,
    unknown: 1,
    matte: 1,
    "grow area": 1,
  };
  if (name && genericNames[name.toLowerCase()]) name = null;

  var hasName = name && name.length > 0;
  var hasValue = value && value.length > 0;
  var shouldSkip = !!skipRoles[role] && !hasName && !hasValue;

  if (!shouldSkip && (hasName || hasValue || interactable) && bbox) {
    elementId++;
    elements.push({
      id: "jxa_" + elementId,
      role: role,
      name: name || null,
      value: value || null,
      bbox: bbox,
      interactable: interactable,
    });
  }

  // Traverse children
  try {
    var children = el.uiElements();
    for (
      var i = 0;
      i < children.length && elements.length < MAX_ELEMENTS;
      i++
    ) {
      traverseElement(children[i], depth + 1, elements);
    }
  } catch (e) {}
}

function main() {
  var result = {
    status: "ok",
    source: "jxa_system_events",
    app_name: null,
    window_title: null,
    elements: [],
  };

  try {
    var se = Application("System Events");
    se.includeStandardAdditions = true;

    var procs = se.processes.whose({ frontmost: true });
    if (procs.length === 0) {
      result.status = "error_no_frontmost_process";
      writeStdout(JSON.stringify(result));
      return;
    }

    var frontProc = procs[0];
    result.app_name = safeGet(function () {
      return frontProc.name();
    });

    var windows = safeGet(function () {
      return frontProc.windows();
    });
    if (windows && windows.length > 0) {
      result.window_title = safeGet(function () {
        return windows[0].name();
      });
      traverseElement(windows[0], 0, result.elements);
    }
  } catch (e) {
    result.status = "error";
  }

  writeStdout(JSON.stringify(result));
}

function writeStdout(str) {
  var data = $.NSString.alloc
    .initWithString(str)
    .dataUsingEncoding($.NSUTF8StringEncoding);
  $.NSFileHandle.fileHandleWithStandardOutput.writeData(data);
}

main();

File: src/screen_probe.swift
#!/usr/bin/env swift
// ZeroClaw Hybrid Programmatic Grounding — Tier 1: Swift AXAPI Probe
// Compiled on first use by screen_state.rs, cached at /tmp/zeroclaw_screen_probe_*
//
// Reads the frontmost application's UI element tree via macOS Accessibility API
// and outputs a JSON ScreenState to stdout.

import Foundation
import ApplicationServices
import AppKit

// MARK: - Data types (match Rust ScreenState schema)

struct BBox: Codable {
    let x: Int
    let y: Int
    let w: Int
    let h: Int
}

struct UIElement: Codable {
    let id: String
    let role: String
    let name: String?
    let value: String?
    let bbox: BBox?
    let interactable: Bool
}

struct ScreenState: Codable {
    let status: String
    let source: String
    let app_name: String?
    let window_title: String?
    let elements: [UIElement]
}

// MARK: - Accessibility helpers

func isTrusted() -> Bool {
    let opts = [kAXTrustedCheckOptionPrompt.takeUnretainedValue() as String: false] as CFDictionary
    return AXIsProcessTrustedWithOptions(opts)
}

func getAttr(_ el: AXUIElement, _ attr: String) -> AnyObject? {
    var value: AnyObject?
    let err = AXUIElementCopyAttributeValue(el, attr as CFString, &value)
    return err == .success ? value : nil
}

func getStringAttr(_ el: AXUIElement, _ attr: String) -> String? {
    guard let val = getAttr(el, attr) else { return nil }
    if let s = val as? String, !s.isEmpty { return s }
    return nil
}

func getBoolAttr(_ el: AXUIElement, _ attr: String) -> Bool? {
    guard let val = getAttr(el, attr) else { return nil }
    if let n = val as? NSNumber { return n.boolValue }
    return nil
}

func getFrame(_ el: AXUIElement) -> BBox? {
    var posVal: AnyObject?
    var sizeVal: AnyObject?
    guard AXUIElementCopyAttributeValue(el, kAXPositionAttribute as CFString, &posVal) == .success,
          AXUIElementCopyAttributeValue(el, kAXSizeAttribute as CFString, &sizeVal) == .success
    else { return nil }

    var point = CGPoint.zero
    var size = CGSize.zero
    guard AXValueGetValue(posVal as! AXValue, .cgPoint, &point),
          AXValueGetValue(sizeVal as! AXValue, .cgSize, &size)
    else { return nil }

    let w = Int(size.width)
    let h = Int(size.height)
    // Skip invisible / zero-size elements
    guard w > 0 && h > 0 else { return nil }

    return BBox(x: Int(point.x), y: Int(point.y), w: w, h: h)
}

// MARK: - Role classification

let interactableRoles: Set<String> = [
    "AXButton", "AXTextField", "AXTextArea", "AXCheckBox",
    "AXRadioButton", "AXPopUpButton", "AXComboBox", "AXSlider",
    "AXLink", "AXMenuItem", "AXMenuBarItem", "AXTab", "AXTabGroup",
    "AXDisclosureTriangle", "AXIncrementor", "AXColorWell",
    "AXToolbar", "AXList", "AXOutline", "AXTable", "AXScrollBar",
    "AXCell", "AXRow", "AXSwitch", "AXToggle"
]

// Roles we skip entirely (container-only, no useful info for the agent)
let skipRoles: Set<String> = [
    "AXGroup", "AXSplitGroup", "AXSplitter", "AXScrollArea",
    "AXLayoutArea", "AXLayoutItem", "AXMatte", "AXRuler",
    "AXRulerMarker", "AXGrowArea", "AXUnknown"
]

// MARK: - Tree traversal

var elementCounter = 0
let maxDepth = 15
let maxElements = 500

func traverse(_ el: AXUIElement, depth: Int, elements: inout [UIElement]) {
    guard depth < maxDepth, elements.count < maxElements else { return }

    let role = getStringAttr(el, kAXRoleAttribute as String) ?? "AXUnknown"

    // Get name from title, description, or help text
    let name = getStringAttr(el, kAXTitleAttribute as String)
        ?? getStringAttr(el, kAXDescriptionAttribute as String)
        ?? getStringAttr(el, kAXHelpAttribute as String)

    // Get value (text content, checkbox state, etc.)
    var value: String? = nil
    if let raw = getAttr(el, kAXValueAttribute as String) {
        if let s = raw as? String, !s.isEmpty {
            // Truncate very long values (e.g. text editor content)
            value = s.count > 200 ? String(s.prefix(197)) + "..." : s
        } else if let n = raw as? NSNumber {
            value = n.stringValue
        }
    }

    let bbox = getFrame(el)
    let interactable = interactableRoles.contains(role)

    // Include element if it has useful info (skip noise)
    let hasName = name != nil
    let hasValue = value != nil
    let shouldSkip = skipRoles.contains(role) && !hasName && !hasValue

    if !shouldSkip && (hasName || hasValue || interactable) && bbox != nil {
        elementCounter += 1
        elements.append(UIElement(
            id: "ax_\(elementCounter)",
            role: role,
            name: name,
            value: value,
            bbox: bbox,
            interactable: interactable
        ))
    }

    // Traverse children
    guard let children = getAttr(el, kAXChildrenAttribute as String) as? [AXUIElement] else { return }
    for child in children {
        traverse(child, depth: depth + 1, elements: &elements)
    }
}

// MARK: - Main

func outputJSON(_ state: ScreenState) {
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.sortedKeys]
    guard let data = try? encoder.encode(state),
          let json = String(data: data, encoding: .utf8)
    else {
        print("{\"status\":\"error_encoding\",\"source\":\"swift_axapi\",\"elements\":[]}")
        return
    }
    print(json)
}

// 1. Check Accessibility permission
guard isTrusted() else {
    outputJSON(ScreenState(
        status: "error_no_accessibility",
        source: "swift_axapi",
        app_name: nil,
        window_title: nil,
        elements: []
    ))
    exit(1)
}

// 2. Get frontmost application
guard let frontApp = NSWorkspace.shared.frontmostApplication else {
    outputJSON(ScreenState(
        status: "error_no_frontmost_app",
        source: "swift_axapi",
        app_name: nil,
        window_title: nil,
        elements: []
    ))
    exit(1)
}

let pid = frontApp.processIdentifier
let appElement = AXUIElementCreateApplication(pid)
let appName = frontApp.localizedName

// 3. Get window title from the first window
var windowTitle: String? = nil
if let windows = getAttr(appElement, kAXWindowsAttribute as String) as? [AXUIElement],
   let firstWindow = windows.first {
    windowTitle = getStringAttr(firstWindow, kAXTitleAttribute as String)
}

// 4. Traverse UI element tree
var elements: [UIElement] = []
traverse(appElement, depth: 0, elements: &elements)

// 5. Output result
outputJSON(ScreenState(
    status: "ok",
    source: "swift_axapi",
    app_name: appName,
    window_title: windowTitle,
    elements: elements
))
